File: sdk-ios/Sources/UXClone/CrashRecorder.swift
import Foundation

typealias PushFn = (UXEvent) -> Void

/// Captures uncaught Objective-C exceptions and routes them as crash events.
final class CrashRecorder {
    private let push:          PushFn
    private let getElapsedMs:  () -> Int64
    private let getCurrentScreen: () -> String
    private let previousHandler: NSUncaughtExceptionHandler?

    init(push: @escaping PushFn, getElapsedMs: @escaping () -> Int64, getCurrentScreen: @escaping () -> String) {
        self.push             = push
        self.getElapsedMs     = getElapsedMs
        self.getCurrentScreen = getCurrentScreen
        self.previousHandler  = NSGetUncaughtExceptionHandler()
        self.attach()
    }

    private func attach() {
        NSSetUncaughtExceptionHandler { [weak self] exception in
            guard let self = self else { return }
            let event = UXEvent(
                type:       .crash,
                elapsedMs:  self.getElapsedMs(),
                screenName: self.getCurrentScreen(),
                value:      String(exception.name.rawValue.prefix(500)),
                metadata:   [
                    "reason":     String(exception.reason?.prefix(500) ?? ""),
                    "error_type": "nsexception",
                ]
            )
            self.push(event)
            // Chain to previous handler
            self.previousHandler?(exception)
        }
    }

    func detach() {
        NSSetUncaughtExceptionHandler(previousHandler)
    }
}

File: sdk-android/library/src/main/kotlin/app/uxclone/sdk/CrashRecorder.kt
package app.uxclone.sdk

/** Captures unhandled JVM exceptions by decorating the default handler. */
internal class CrashRecorder(
    private val push:             (UXEvent) -> Unit,
    private val getElapsedMs:     () -> Long,
    private val getCurrentScreen: () -> String,
) {
    private val previousHandler = Thread.getDefaultUncaughtExceptionHandler()

    init { attach() }

    private fun attach() {
        Thread.setDefaultUncaughtExceptionHandler { thread, throwable ->
            val event = UXEvent(
                type       = UXEventType.CRASH,
                elapsedMs  = getElapsedMs(),
                screenName = getCurrentScreen(),
                value      = throwable.message?.take(500) ?: "Unknown error",
                metadata   = mapOf(
                    "stack"      to (throwable.stackTraceToString().take(2000)),
                    "thread"     to thread.name,
                    "error_type" to "uncaught_exception",
                )
            )
            push(event)
            previousHandler?.uncaughtException(thread, throwable)
        }
    }

    fun detach() {
        Thread.setDefaultUncaughtExceptionHandler(previousHandler)
    }
}
